### ISSUE TITLE
Support coverage in Mojo::Template

### ISSUE THREAD

**Issue by ksmadsen** [kraih/mojo#960]
As requested in #956, this is an issue to get `Devel::Cover` support for `Mojo::Template`.

Since the original pull request, I've done some further experiments. The following patch is enough to make `Devel::Cover` give correct coverage for template files:
```perl
diff --git a/lib/Mojo/Template.pm b/lib/Mojo/Template.pm
index 8f17b40..f0a7382 100644
--- a/lib/Mojo/Template.pm
+++ b/lib/Mojo/Template.pm
@@ -167,6 +167,9 @@ sub process {
     return Mojo::Exception->new($@)->inspect($self->unparsed, $code)
       ->trace->verbose(1)
       unless $compiled = eval $self->_wrap($code, @_);
+    no strict 'refs';
+    *{$self->namespace . "::" . Mojo::Util::md5_sum $self} = $compiled;
+    use strict 'refs';
     $self->compiled($compiled);
   }
 
@@ -186,7 +189,7 @@ sub render { shift->parse(shift)->process(@_) }
 sub render_file {
   my ($self, $path) = (shift, shift);
 
-  $self->name($path) unless defined $self->{name};
+  $self->name($path);
   my $template = slurp $path;
   my $encoding = $self->encoding;
   croak qq{Template "$path" has invalid encoding}
```
This is just for illustration, I'm aware that there are quite a few issues with the patch.

Using the above patch, the sub name in exceptions is still `__ANON__`, so the sub isn't named.

So basically, for coverage to work two things are needed:

1. Correct `#line` directives in the template code (e.g. referencing the correct line number and file name).
2. The anonymous sub that the template is compiled into, needs to be accessible from the symbol table upon exit.

If injecting the sub into the symbol table is a performance/memory issue, this could be dependent on an environment variable as suggested by @jberger.

**Comment by kraih** [comment]
I know you said that you're aware of the shortcomings of the example patch, but i think it's worth pointing out that the whole approach doesn't work because it results in a serious memory leak.

**Comment by ksmadsen** [comment]
Yes, I can see how my patch will result in a memory leak.

However, the anonymous sub needs to be available when the program exits for `Devel::Cover` to see it, so there is no way around leaking it.

That basically leaves the `MOJO_TEMPLATE_COVER` solution as the only viable option, if this feature is desired?

**Comment by kraih** [comment]
If a memory leak is required, then i'm not sure we should be doing this at all.

**Comment by kraih** [comment]
An env variable like `MOJO_TEMPLATE_COVER` results in a lot of technical debt, i just don't see `Devel::Cover` support being worth it.

**Comment by ksmadsen** [comment]
For our part, I think that this is one of our most wanted features, that is missing from Mojolicious.

If it isn't the most wanted feature, it is at least the feature that we've spent the most time discussing how to implement. And our previous best suggestion (generating template code, save it to a file, and then use it), is even more of a kludge.

FWIW, I've prepared a patch to make EP/EPL templates use the file path instead of the name: af9743db3b0f76e809b863f2f5a0393b72110003
I've also prepared a patch which makes the sub accessible from the symbol table if `MOJO_TEMPLATE_COVER` is set: 067576283827db0bc96a8d6e82cacf1609cb05d3

**Comment by kraih** [comment]
For now i'm afraid i'll have to give this feature a 👎 vote, because there's no clean way to implement it.

**Comment by kraih** [comment]
Something that has not been mentioned yet is support for templates in `DATA` sections and inline templates, which i think are important parts for making this an officially supported feature.

**Comment by Eckankar** [comment]
I feel that this feature could bring tremendous value even with only being supported in file templates. (And a disclaimer could be added in the documentation, saying that it's only supported on those.)

We have the use-case, that we have a rather large codebase, and would like to make sure that our testcases exercise as much of our code as is possible. Using Devel::Cover, we can get a pretty good idea if we've covered all the code-paths in our testing.

However, since we cannot get coverage information on templates, they can easily not be fully covered by our tests without us knowing. This relatively often comes back to bite us, when the some un-covered part turns out to produce some sort of error when it is run. A large part of these sorts of bugs can be avoided more easily with the ability to see coverage. (Hence why we've been discussing it back and forth in the office for quite a while.)

**Comment by kraih** [comment]
There are many other solutions worth exploring. Like changing the names assigned in `Mojolicious::Plugin::EPLRenderer` a little, to make them more suitable for these kinds of use cases, and then having a hook somewhere that allows for the monkey patching hack to be added on demand.

**Comment by kraih** [comment]
Actually, i bet it's already possible with a clever `prepend` value.
```
my $prepend = 'no strict "refs"; *{"Mojo::Template::Sandbox::" . Mojo::Util::md5_sum __FILE__} = __SUB__;';
plugin EPRenderer => {name => 'ep', template => {prepend => $prepend}};
```

**Comment by ksmadsen** [comment]
> Something that has not been mentioned yet is support for templates in DATA sections and inline templates, which i think are important parts of making this an officially supported feature.

How would you find the originating line for an inline template? It quite hard to come up with a correct originating line-number for code like:

```perl
$c->render(inline => EOF);
Content
ABC
EOF

$c->render(inline => 'Content\nABC\n');
````
And completely impossible for string variables.

I believe that it is feasible for `__DATA__` templates. However, since the line numbers needs to be correct, it also means that error messages will be referencing the line number from the originating file, instead of the line number from the start of the template. So `Mojo::Exception` will also have know about this.

**Comment by ksmadsen** [comment]
> Actually, i bet it's already possible with a clever prepend value.

That actually works, if af9743db3b0f76e809b863f2f5a0393b72110003 is also applied (and then only for file based templates).

I was afraid that it would hit pjcj/Devel--Cover#72, but it seems that the code needs to call the `__SUB__`, in order to trigger that bug.

**Comment by kraih** [comment]
@ksmadsen You'll have to come up with a new consistent naming scheme for log messages then, because that patch messes things up quite a bit.
```
[Fri May 13 17:51:55 2016] [debug] Rendering cached /Users/sri/repo/mojo/lib/Mojo/../Mojolicious/resources/templates/mojo/debug.html.ep
...
[Fri May 13 17:53:19 2016] [debug] Rendering template "foo/bar.rss.ep"
...
[Fri May 13 17:53:19 2016] [debug] Rendering cached template "foo/fun.html.ep" from DATA section
```

**Comment by ksmadsen** [comment]
Instead of af9743db3b0f76e809b863f2f5a0393b72110003, the following gives the right filenames in the line directive and the right template name in the debug messages:

```perl
diff --git a/lib/Mojolicious/Plugin/EPLRenderer.pm b/lib/Mojolicious/Plugin/EPLRenderer.pm
index 9d8df28..3c5fd41 100644
--- a/lib/Mojolicious/Plugin/EPLRenderer.pm
+++ b/lib/Mojolicious/Plugin/EPLRenderer.pm
@@ -38,7 +38,8 @@ sub _render {
       # Try template
       if (defined(my $path = $renderer->template_path($options))) {
         $c->app->log->debug(qq{Rendering template "$name"});
-        $$output = $mt->name(qq{template "$name"})->render_file($path, @args);
+        $$output = $mt->render_file($path, @args);
+        $mt->name(qq{template "$name"});
       }
 
       # Try DATA section
```

The patch for `lite_app.t` from af9743db3b0f76e809b863f2f5a0393b72110003 is still needed to make the tests pass.

**Comment by ksmadsen** [comment]
Crap... Too quick. That one doesn't work if the template doesn't compile in the first try.

Then I see no other solution than having a path attribute in `Mojo::Template`? Passing the path/name around within the functions in `Mojo::Template` seems like it would complicate things unnecessarily?

**Comment by kraih** [comment]
@ksmadsen Please stop with the hacks, i want a clean solution like a new naming scheme.

**Comment by jberger** [comment]
> @ksmadsen Please stop with the hacks, i want a clean solution like a new naming scheme.

Said differently, it is ok if we look at rearchitecting a few things here.

**Comment by kraih** [comment]
Looks like this issue went stale.
